ceph/rbd_mirror_status.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// RbdMirrorStatusCollector displays statistics about each pool in the Ceph cluster.
+type RbdMirrorStatusCollector struct {
+	config  string
+	logger  *logrus.Logger
+	version *Version
+
+	getRbdMirrorStatus func(config string) ([]byte, error)
+
+	// RbdMirrorStatus shows the overall health status of a rbd-mirror.
+	RbdMirrorStatus prometheus.Gauge
+
+	// RbdMirrorDaemonStatus shows the health status of a rbd-mirror daemons.
+	RbdMirrorDaemonStatus prometheus.Gauge
+
+	// RbdMirrorImageStatus shows the health status of rbd-mirror images.
+	RbdMirrorImageStatus prometheus.Gauge
+}
+
+// rbdMirrorStatus get the RBD Mirror Pool Status
+var rbdMirrorStatus = func(config string) ([]byte, error) {
+	out, err := exec.Command(rbdPath, "-c", config, "mirror", "pool", "status", "--format", "json").Output()
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// NewRbdMirrorStatusCollector creates a new RbdMirrorStatusCollector instance
+func NewRbdMirrorStatusCollector(exporter *Exporter) *RbdMirrorStatusCollector {
+	labels := make(prometheus.Labels)
+	labels["cluster"] = exporter.Cluster
+
+	collector := &RbdMirrorStatusCollector{
+		config:  exporter.Config,
+		logger:  exporter.Logger,
+		version: exporter.Version,
+
+		getRbdMirrorStatus: rbdMirrorStatus,
+
+		RbdMirrorStatus: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   cephNamespace,
+				Name:        "rbd_mirror_pool_status",
+				Help:        "Health status of rbd-mirror, can vary only between 3 states (err:2, warn:1, ok:0)",
+				ConstLabels: labels,
+			},
+		),
+
+		RbdMirrorDaemonStatus: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   cephNamespace,
+				Name:        "rbd_mirror_pool_daemon_status",
+				Help:        "Health status of rbd-mirror daemons, can vary only between 3 states (err:2, warn:1, ok:0)",
+				ConstLabels: labels,
+			},
+		),
+
+		RbdMirrorImageStatus: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Namespace:   cephNamespace,
+				Name:        "rbd_mirror_pool_image_status",
+				Help:        "Health status of rbd-mirror images, can vary only between 3 states (err:2, warn:1, ok:0)",
+				ConstLabels: labels,
+			},
+		),
+	}
+
+	return collector
+}
+
+func (c *RbdMirrorStatusCollector) metricsList() []prometheus.Metric {
+	return []prometheus.Metric{
+		c.RbdMirrorStatus,
+		c.RbdMirrorDaemonStatus,
+		c.RbdMirrorImageStatus,
+	}
+}
+
+// Describe provides the metrics descriptions to Prometheus
+func (c *RbdMirrorStatusCollector) Describe(ch chan<- *prometheus.Desc) {
+	for _, metric := range c.metricsList() {
+		ch <- metric.Desc()
+	}
+}
+
+// Collect sends all the collected metrics Prometheus.
+func (c *RbdMirrorStatusCollector) Collect(ch chan<- prometheus.Metric) {
+	status, err := rbdMirrorStatus(c.config)
+	if err != nil {
+		c.logger.WithError(err).Error("failed to run 'rbd mirror pool status'")
+	}
+	var rbdStatus rbdMirrorPoolStatus
+	if err = json.Unmarshal(status, &rbdStatus); err != nil {
+		//blabla
```

Comment:
https://www.youtube.com/watch?v=Hrph2EW9VjY

Reply:
oops

Reply:
I have that album!